azurerm/internal/services/loganalytics/parse/log_analytics_storage_insights.go

```diff
@@ -0,0 +1,61 @@
+package parse
+
+import (
+	"fmt"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+)
+
+type LogAnalyticsStorageInsightsId struct {
+	ResourceGroup string
+	WorkspaceName string
+	WorkspaceID   string
+	Name          string
+}
+
+// Note - this API currently lower-cases all return values
+// Issue tracked here - https://github.com/Azure/azure-sdk-for-go/issues/13268
+const fmtWorkspaceId = "/subscriptions/%s/resourcegroups/%s/providers/%s/workspaces/%s"
```

Comment:
let's make this it's own type with it's own object:

```suggestion
type LogAnalyticsWorkspaceId struct {
    ResourceGroup string
    Name string
}

func (id LogAnalyticsWorkspaceId) ID(subscriptionId string) string {
    fmtString := "/subscriptions/%s/resourceGroups/%s/providers/%s/workspaces/%s"
    return fmt.Sprintf(fmtString, subscriptionId, id.ResourceGroup, id.Name)
}

func NewLogAnalyticsWorkspaceID(resourceGroup, name string) LogAnalyticsWorkspaceId {
    return LogAnalyticsWorkspaceId{
      Name: name,
      ResourceGroup: resourceGroup,
    }
}
```

which then means we can embed the "LogAnalyticsWorkspaceId" object within the "LogAnalyticsStorageInsight" struct below, and call the nested "ID" function to build up this ID (and use it within the Read function to ensure it's consistent)

We probably could (separately) add this to the Log Analytics Workspace resource and add a state migration to ensure these are consistent?

Reply:
my bad - half a job, this already exists, I'll refactor to use it.